index.rst

```diff
@@ -13,17 +13,542 @@
+DMTN-249 :cite:`DMTN-249` attempts to address this issue by defining a new consistency model that explicitly permits (and actually requires) file artifacts to exist prior to their datasets' inclusion in the registry, as long as those datasets are tracked as *possibly* existing in some other way that allows them to be found by any client of the central data repository (not merely, e.g. one that knows the user-provided location of an external quantum graph file, as is the case today).
+The workspace concept described here is a realization of this new consistency model: the central data repository will have a record of all active (internal) workspaces (probably in a registry database table) and its full butler clients will have the ability to construct a :class:`WorkspaceButler` for these by name
+Workspace implementations are required to support the :meth:`WorkspaceButler.abandon` operation, which effectively requires that it have a way to find all of its file artifacts.
+
+Concurrency and workspace consistency
+-------------------------------------
+
+Different workspaces belonging to the same central data repository are fully independent, and may be operated on concurrently in different processes with no danger of corruption.
+This includes commits and workspace creation, and races for creation of the same workspace should result in at least one of the two attempts explicitly failing, and in no cases should hardware failures or concurrency yield an incorrectly-created workspace that can masquerade as a correct one.
+
+Operations on workspaces should be idempotent either by being atomic or by being resumable from a partial state after interruption, with exceptions to this rule clearly documented.
+
+*Most* operations on a workspaces are not expected to support or guard against concurrent operations on the same workspace, unless they are operating on disjoint sharding data IDs, in which case concurrency is required of all sharding workspace implementations.
+In addition, some workspace-specific interfaces are expected to explicitly support concurrency under certain conditions; this is, after all, much of the reason we require workspaces to avoid central database access whenever possible.
+Most prominent of these is the :meth:`PipelineWorkspaceButler.run_quanta` method, which guarantees correct behavior in the presence of concurrent calls as long as those calls identify disjoint sets of quanta to run, allowing it to be used as an execution primitive by higher-level logic capable of generating those disjoint sets (e.g. BPS).
+
+In fact, we will define :meth:`PipelineWorkspaceButler.run_quanta` to *block* when it is given a quantum whose inputs are produced by other quanta that have not been run.
+This is possible because this workspace implementation will also record quantum status (e.g. in per-quantum report files) for provenance, and we can modify the single-quantum execution logic to wait (e.g. for these files) to appear upstream.
```

Comment:
I am a bit worried about using files on filesystem for inter-process communication. Filesystems are not super-reliable medium, maybe some messaging protocol could work better here. And for single-process case (like single `pipetask` running many quanta) we can just pass that info from executor to caller.

Reply:
I completely agree, but it's effectively what we're doing now (except we're making sketchy assumptions about what file existence means rather than blocking).  And every time I ask on Slack about, "oh, what about using scalable NoSQL database X", everyone asks if I'm _really_ certain it's the best one for the job while I'm just thinking, "gah, anything is better than files for this!" :slightly_smiling_face: .

In the single-process and multiprocessing case I do imagine us passing that information around in-memory (or whatever pipes/sockets multiprocessing does under the hood), but I _also_ think we need to write the information to files for checkpointing - though I'm not sure if that has to be done once per quantum or once per `run_quanta` call (which translates to once per BPS job i.e. cluster in batch).